smart_open/http.py

```diff
@@ -114,3 +133,118 @@ def readinto(self, b):
             return 0
         b[:len(data)] = data
         return len(data)
+
+
+class SeekableBufferedInputBase(BufferedInputBase):
+    """
+    Implement seekable streamed reader from a web site.
+    Supports Kerberos and Basic HTTP authentication.
+    """
+
+    def __init__(self, url, mode='r', buffer_size=DEFAULT_BUFFER_SIZE,
+                 kerberos=False, user=None, password=None):
+        """
+        If Kerberos is True, will attempt to use the local Kerberos credentials.
+        Otherwise, will try to use "basic" HTTP authentication via username/password.
+
+        If none of those are set, will connect unauthenticated.
+        """
+        self.url = url
+
+        if kerberos:
+            import requests_kerberos
+            self.auth = requests_kerberos.HTTPKerberosAuth()
+        elif user is not None and password is not None:
+            self.auth = (user, password)
+        else:
+            self.auth = None
+
+        self.buffer_size = buffer_size
+        self.mode = mode
+        self.response = self._partial_request()
+
+        if not self.response.ok:
+            self.response.raise_for_status()
+
+        logger.debug('self.response: %r, raw: %r', self.response, self.response.raw)
+
+        self._seekable = True
+
+        self.content_length = int(self.response.headers.get("Content-Length", -1))
+        if self.content_length < 0:
+            self._seekable = False
+        if self.response.headers.get("Accept-Ranges", "none").lower() != "bytes":
+            self._seekable = False
+
+        self._read_iter = self.response.iter_content(self.buffer_size)
+        self._read_buffer = b''
+        self._current_pos = 0
+
+        #
+        # This member is part of the io.BufferedIOBase interface.
+        #
+        self.raw = None
+
+    def seek(self, offset, whence=0):
+        """Seek to the specified position.
+
+        :param int offset: The offset in bytes.
+        :param int whence: Where the offset is from.
+
+        Returns the position after seeking."""
+        logger.debug('seeking to offset: %r whence: %r', offset, whence)
+        if whence not in WHENCE_CHOICES:
+            raise ValueError('invalid whence, expected one of %r' % WHENCE_CHOICES)
+
+        if not self.seekable():
+            raise OSError
+
+        if whence == START:
+            new_pos = offset
+        elif whence == CURRENT:
+            new_pos = self._current_pos + offset
+        elif whence == 2:
+            new_pos = self.content_length + offset
+
+        new_pos = _clamp(new_pos, 0, self.content_length)
+
+        if self._current_pos != new_pos:
```

Comment:
I'd use an early return here. Saves the reader from remembering they're inside a conditional, and saves the writer an extra level of indentation. Also dodges the need to nest conditionals.

```python
if self._current_pos == new_pos:
    return self._current_pos
```